File: src/livingwp/utils/markdown.py
import frontmatter


def parse_markdown(text: str) -> tuple[dict[str, object], str]:
    """Parse a markdown file with YAML front matter.

    Returns a tuple of (front_matter_dict, body).
    """
    return frontmatter.parse(text)


def format_markdown(metadata: dict[str, object], body: str) -> str:
    """Combines metadata and body into markdown document with YAML frontmatter.

    Returns the thge text of the markdown document with frontmatter
    """
    post = frontmatter.Post(body, **metadata)
    return frontmatter.dumps(post)

File: src/agent/src/markdown.ts
import matter from 'gray-matter';
import { dump, type DumpOptions } from 'js-yaml';

export type FrontMatter = Record<string, unknown>;

/** Parse a markdown document with YAML frontmatter into (metadata, body). */
export function parseMarkdown(text: string): [FrontMatter, string] {
  const { data, content } = matter(text);
  return [data, content.replace(/^\n/, '')];
}

/**
 * Combine metadata and body into a markdown document with YAML frontmatter.
 *
 * Matches the output conventions of python-frontmatter, which wrote all
 * existing content: alphabetically sorted keys, ~80 column wrapping, and no
 * trailing newline after the body.
 */
export function formatMarkdown(metadata: FrontMatter, body: string): string {
  const frontMatter = dump(metadata, {
    sortKeys: true,
    lineWidth: 80,
    allowUnicode: true,
  } as DumpOptions);
  return `---\n${frontMatter}---\n\n${body.trim()}`;
}
